### DISCUSSION TITLE
PVoutput.org Tool

### DISCUSSION THREAD

**Discussion by jasonacox** [jasonacox/Powerwall-Dashboard#46] [Announcements]
Several in the community publish their solar production data to [PVoutput.org](https://pvoutput.org/), a free service for publicly sharing and comparing PV output data.  Since this Powerwall-Dashboard project stores all energy production data, it is relatively easy to pull this out of the dashboard and publish it to PVoutput on a one-time or regular basis.

I have created a python script, [pvoutput.py](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/tools/pvoutput.py) to publish Powerwall data to PVoutput.  The script is in the [tools](https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools) directory for this project.

To use the script:
* Sign up at [pvoutput.org](https://pvoutput.org/account.jsp) to get an API KEY - update the settings in the script with your API_SYSTEM_ID and API_KEY.
* Update the INFLUXDB_HOST in the script to the address of your Dashboard host (default = localhost) and INFLUXDB_TZ to your timezone.
* Run the script:

```bash
python3 pvoutput.py 
```
It will run an interactive mode:

```
Select Custom Date Range
 - Enter start day (YYYY-mm-dd): 2022-06-20
 - Enter end date (YYYY-mm-dd): 2022-06-26

Sending Solar Data [2022-06-20 to 2022-06-27]
2022-06-20:    Generated = 49256 - Exported = 9067 - Consumed = 43501 - Imported = 2455 - Published
2022-06-21:    Generated = 49543 - Exported = 8386 - Consumed = 47026 - Imported = 7581 - Published
2022-06-22:    Generated = 39698 - Exported = 243 - Consumed = 46923 - Imported = 9212 - Published
2022-06-23:    Generated = 47508 - Exported = 127 - Consumed = 58052 - Imported = 12342 - Published
2022-06-24:    Generated = 48716 - Exported = 4474 - Consumed = 51203 - Imported = 8881 - Published
2022-06-25:    Generated = 49506 - Exported = 5609 - Consumed = 51118 - Imported = 9041 - Published
2022-06-26:    Generated = 48904 - Exported = 1083 - Consumed = 54617 - Imported = 8280 - Published
Done.
```
You can also add it to a daily cronjob by specifying a optional parameter:  'yesterday' or 'today'. 

```bash
# Pull and publish yesterdays data
python3 pvoutput.py yesterday

# Pull and publish todays data so far
python3 pvoutput.py today
```

I used this script to import of all the data since my system was installed.  Keep in mind that PVoutput has a rate limit of 60 updates per hour so be careful.  If you are only updating once a day, this shouldn't be a problem.  I donated to help their cause which also increase my rate limit.  That was useful while I developed this script.

You can see my published data here:  https://pvoutput.org/aggregate.jsp?p=0&id=104564&sid=91747&t=m&v=1&s=1 

<img width="960" alt="image" src="https://user-images.githubusercontent.com/836718/175867308-416584ba-82e5-4da8-9cdc-4ece163e1ca2.png">

**Comment by cwagz** [comment]
I am getting an error on line 28. Looks like it can't pull the module.

```
Traceback (most recent call last):
  File "/home/chad/Powerwall-Dashboard/tools/pvoutput.py", line 28, in <module>
    from influxdb import InfluxDBClient
ModuleNotFoundError: No module named 'influxdb'
chad@powerwall:~/Powerwall-Dashboard/tools$
```

**Reply by jasonacox** [reply]
Thanks, @cwagz !  Yes, it does need that module.  I'll add a note or make a code change to prompt the user to install it.

```bash
pip install influxdb

# or

python3 -m pip install influxdb
```

**Reply by cwagz** [reply]
Thanks for the help. I had no idea it was something I needed to install. I ended up having to install pip first as well. Got it working now.

**Reply by cwagz** [reply]
https://pvoutput.org/aggregate.jsp?id=104740&sid=91879&v=1&t=m

**Reply by jasonacox** [reply]
Awesome!  Wow, great production!  I set up a cronjob on my system to send daily updates at 12:07am. 

```bash
# m h  dom mon dow   command
7 0 * * * /home/tesla/pvoutput.py yesterday > /dev/null 2>&1
```

**Reply by cwagz** [reply]
Hi @jasonacox, 

With version 2.5 would we now be able to send "live" data to pvoutput.org with weather data included? I had already signed up for an openweathermap api key, but it didn't seem very useful when only sending data once per day.

Thanks for all the work on this project!

**Reply by jasonacox** [reply]
Yes!  I didn't even think of that. PVOutput allows us to send temp_min (`tm`) and temp_max (`tx`) data for each day. 

The new [pvoutput.py](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/tools/pvoutput.py) script now pulls the min/max temp data from InfluxDB (since we are storing that now) and pushes that to PVOutput. 

I just tested by resending all my August data which includes several days of temp data now (uploaded [here](https://pvoutput.org/list.jsp?id=104564&sid=91747)). Let me know if it works for you!

**Reply by cwagz** [reply]
@jasonacox - awesome! I will give this a try soon.

I assume the script would need more work to be able to send hourly or 15-min increment data? I setup my TOU in pvOutput but it would require "live" data so that it can track savings as the rates change. I see you have some mention of TOU in the script already.

If I have time, I might try tinkering around with it. It would be a good challenge for me.

**Reply by pgatty** [reply]
> Awesome! Wow, great production! I set up a cronjob on my system to send daily updates at 12:07am.
> 
> ```shell
> # m h  dom mon dow   command
> 7 0 * * * /home/tesla/pvoutput.py yesterday > /dev/null 2>&1
> ```

@jasonacox - Sorry for necroposting, but I'm trying to get your pvoutput.py script running from root's cron, but it's not working. I have updated the pvoutput.py script with my API key and other necessary values. I can run the script manually via `./pvoutput.py yesterday` and it works as expected, but it doesn't run from cron. Here's my cron entry created with `sudo crontab -e`

`10 12 * * * /home/pgatty/Powerwall-Dashboard/tools/pvoutput/pvoutput.py yesterday > /dev/null 2>&1`

I've following these instructions for setting up the python environment.

`python3 -m venv .venv`
`source .venv/bin/activate`
`pip install influxdb`

Do this so it always uses that venv
 `echo "source .venv/bin/activate" >> ~/.bashrc`

When I pipe the cron job output to a file, I get this message 
`ERROR: Missing python influxdb module. Run 'pip install influxdb'.`

I've previously run 'pip install influxdb` as root. How do I get the cron job to use it?

**Reply by pgatty** [reply]
I figured it out. You need to use the python binary of your venv in your cron. So this is what my cron job look like now.

`0 5 * * * /home/pgatty/.venv/bin/python /home/pgatty/Powerwall-Dashboard/tools/pvoutput/pvoutput.py yesterday > /dev/null 2>&1`

**Comment by cwagz** [comment]
I was looking at some other python scripts and yours. It seems pretty tricky to do live data with a battery in the mix. There does not seem to be a good way, in pvoutput, to account for all the scenarios that can occur with where the power is coming from or going to. This is probably pretty similar to trying to calculate TOU savings on the dashboard.

![image](https://user-images.githubusercontent.com/13752647/187100919-a24b8256-6f54-4050-b108-ef8b798e501e.png)

**Reply by jasonacox** [reply]
Yes, I'm fairly certain that is why I gave up on making my script do accumulated updates instead of daily.  I believe it assumes energy generation beyond consumption is exported back to the grid.  Daily allows you to explicitly report how much is generated, imported, exported, and consumed.  It may be interesting to see if anyone has solved how to do the updates with an ESS.

Speaking of which, I need to work on better savings calc for the dashboard that does include TOU rate differences. :)

**Comment by billraff** [comment]
I just started using this tool a week ago. It's starting to get hot in Texas and when my job ran to post the daily update to pvoutput last night I received this error - 

Sending Solar Data [2025-05-14]
2025-05-14:    Generated = 42328 - Exported = 75 - Consumed = 44631 - Imported = 69 - Temp = 71.0 / 100.4Traceback (most recent call last):
  File "/volume1/homes/billr/Powerwall-Dashboard/tools/pvoutput/pvoutput.py", line 213, in &lt;module&gt;
    push_daily(day, generated, exported, consumed, imported, tm, tx)
  File "/volume1/homes/billr/Powerwall-Dashboard/tools/pvoutput/pvoutput.py", line 110, in push_daily
    raise ValueError(response.read())
ValueError: b'Bad request 400: Max temp [100.4] greater than 100 degrees celsius on 20250514'

Why is it reporting in celsius?I have my Powerwall-Dashboard set for Imperial for measurements and my pvoutput is set to use Fahrenheit.

<img width="412" alt="Screenshot 2025-05-15 at 7 05 44 AM" src="https://github.com/user-attachments/assets/f6a17769-acf7-4ec4-b515-a106346dfe14" />
<img width="567" alt="Screenshot 2025-05-15 at 7 03 26 AM" src="https://github.com/user-attachments/assets/0cbf53eb-efde-4f44-84c6-b660ae661b9d" />

**Reply by jasonacox** [reply]
You need to adjust the script. It is set to use metric values, see this line:

https://github.com/jasonacox/Powerwall-Dashboard/blob/13a5d98934a6730c1e7664ff1dee82715a0d9cd8/tools/pvoutput/pvoutput.py#L47-L48

change it to `imperial` for your setup.

**Comment by billraff** [comment]
Jason, as always, I appreciate your help.
